std/experimental/checkedint.d

```diff
@@ -0,0 +1,1584 @@
+efficiency. For example, $(D Checked!(uint, ProperCompare)(1) > -1) is `true`,
+which is not the case with the built-in comparison. Also, comparing numbers for
+equality with floating-point numbers only passes if the integral can be
+converted to the floating-point number precisely, so as to preserve transitivity
+of equality.)
+
+$(LI `WithNaN` reserves a special "Not a Number" value. )
+
+)
+
+These policies may be used alone, e.g. $(D Checked!(uint, WithNaN)) defines a
+`uint`-like type that reaches a stable NaN state for all erroneous operations.
+They may also be "stacked" on top of each other, owing to the property that a
+checked integral emulates an actual integral, which means another checked
+integral can be built on top of it. Some interesting combinations include:
+
+$(UL
+
+$(LI $(D Checked!(Checked!int, ProperCompare)) defines an `int` with fixed
+comparison operators that will fail with `assert(0)` upon overflow. (Recall that
+`Croak` is the default policy.) The order in which policies are combined is
+important because the outermost policy (`ProperCompare` in this case) has the
+first crack at intercepting an operator. The converse combination $(D
+Checked!(Checked!(int, ProperCompare))) is meaningless because `Croak` will
+intercept comparison and will fail without giving `ProperCompare` a chance to
+intervene.)
+
+$(LI $(D Checked!(Checked!(int, ProperCompare), WithNaN)) defines an `int`-like
+type that supports a NaN value. For values that are not NaN, comparison works
+properly. Again the composition order is important; $(D Checked!(Checked!(int,
+WithNaN), ProperCompare)) does not have good semantics because `ProperCompare`
+intercepts comparisons before the numbers involved are tested for NaN.)
+
+)
+
+*/
+module std.experimental.checkedint;
+import std.traits : isFloatingPoint, isIntegral, isNumeric, isUnsigned, Unqual;
+import std.conv : unsigned;
+
+/**
+Checked integral type wraps an integral `T` and customizes its behavior with the
+help of a `Hook` type. The type wrapped must be one of the predefined integrals
+(unqualified), or another instance of `Checked`.
+*/
+struct Checked(T, Hook = Croak)
+if (isIntegral!T && is(T == Unqual!T) || is(T == Checked!(U, H), U, H))
+{
+    import std.algorithm : among;
+    import std.traits : hasMember;
+    import std.experimental.allocator.common : stateSize;
+
+    /**
+    The type of the integral subject to checking.
+    */
+    alias Payload = T;
+
+    // state {
+    static if (hasMember!(Hook, "defaultValue"))
+        private T payload = Hook.defaultValue!T;
+    else
+        private T payload;
+    /**
+    `hook` is a member variable if it has state, or an alias for `Hook`
+    otherwise.
+    */
+    static if (stateSize!Hook > 0) Hook hook;
+    else alias hook = Hook;
+    // } state
+
+    // representation
+    /**
+    Returns a copy of the underlying value.
+    */
+    auto representation() inout { return payload; }
```

Comment:
std.typecons.Nullable uses `get` for this function. Might make sense to use the same name